Implementing the controllers

lianet89 opened the pull request:
Implementing in the classes the behaviors described in the excersice.

lianet89 committed:
Implementing the controllers

Implementing the methods in the controllers classes.

andrydiaz commented on src/main/java/com/example/springBoot_hibernate_ferrocarriles/controller/LocomotoraController.java:
```diff
@@ -0,0 +1,53 @@
+	@Autowired
+	LocomotoraService locomotoraService;
+
+	public LocomotoraController() {
```
No  necesitas el constructor

andrydiaz commented on src/main/java/com/example/springBoot_hibernate_ferrocarriles/controller/LocomotoraController.java:
```diff
@@ -0,0 +1,53 @@
+		super();
+	}
+	
+	@GetMapping("/locomotora")
```
Como es una lista es mejor `/locomotoras`, con `s` :)

andrydiaz commented on src/main/java/com/example/springBoot_hibernate_ferrocarriles/controller/LocomotoraController.java:
```diff
@@ -0,0 +1,53 @@
+		return locomotoraService.getAllLocomotora();
+	}
+	
+	@GetMapping("/locomotora/{locomotoraId}")
```
En esta url, podemos eliminar redundancia: `/locomotora/{id}`

andrydiaz commented on src/main/java/com/example/springBoot_hibernate_ferrocarriles/controller/LocomotoraController.java:
```diff
@@ -0,0 +1,53 @@
+	}
+	
+	@PostMapping("/locomotora")
+	public int addLocomotora(@RequestBody Locomotora locomotora){
```
Consider to return ResponseEntity object instead of primitives types(boolean, int, etc).
The method `addOrUpdateLocomotora` in `LocomotoraService` could return a Locomotora entity, hence the controller too.

andrydiaz commented on src/main/java/com/example/springBoot_hibernate_ferrocarriles/controller/LocomotoraController.java:
```diff
@@ -0,0 +1,53 @@
+	}
+	
+	@PutMapping("/locomotora")
+	public Locomotora updateLocomotora(@RequestBody Locomotora locomotora){
```
If the update returns a Locomotora why not the add?

andrydiaz commented on src/main/java/com/example/springBoot_hibernate_ferrocarriles/controller/CocheMotorController.java:
```diff
@@ -0,0 +1,54 @@
+		super();
+	}
+	
+	@GetMapping("/cocheMotor")
```
Just a suggestion, URL are not necessarily following Java conventions, it is ok to have: `cochemotor`.

andrydiaz commented on src/main/java/com/example/springBoot_hibernate_ferrocarriles/controller/EquipoDeTraccionController.java:
```diff
@@ -0,0 +1,55 @@
+	@Autowired
+	EquipoDeTraccionService equipoDeTraccionService;
+
+	public EquipoDeTraccionController() {
```
Not need the constructor

andrydiaz commented on src/main/java/com/example/springBoot_hibernate_ferrocarriles/controller/ReportesController.java:
```diff
@@ -0,0 +1,94 @@
+	@GetMapping("/reportes/locomotorasDeMarca/{marca}")
+	public List<Locomotora> locomotorasDeMarca(@PathVariable("marca")String marca){
+		List<Locomotora> locomotorasDeMarca = new ArrayList<>();
+		for(Locomotora locomotora:locomotoraService.getAllLocomotora()) {
```
Controller classes shouldn't have logic inside, they rely on Services to perform business logic. So you should have a ReportService class to delegate these functionalities.

andrydiaz commented on src/main/java/com/example/springBoot_hibernate_ferrocarriles/controller/ReportesController.java:
```diff
@@ -0,0 +1,94 @@
+	
+	@GetMapping("/reportes/origenMasSalidas")
+	public String origenMasSalidas() {
+		List<Itinerario> listaItinerarios = itinerarioService.getAllItinerarios();
```
The whole method looks complex, try to separate the method into several methods with small responsibilities. This should be implemented in a service.

andrydiaz commented on src/main/java/com/example/springBoot_hibernate_ferrocarriles/controller/ViajeController.java:
```diff
@@ -0,0 +1,86 @@
+	
+	@PostMapping("/viaje")
+	public Viaje addViaje(@RequestBody Itinerario itinerario){		
+		int numeroViaje = viajeService.getAllViajes().size()+1;
```
`ViajeService` classes should contain this logic, and release the controllers from computing stuff or doing business logic.

andrydiaz commented on src/main/java/com/example/springBoot_hibernate_ferrocarriles/model/CocheMotor.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public int cantidadMaxCoches() {
+        int i=0;
```
A more clear approach 😄 
`
if (potenciaMotor<4000){
    return 1;
} 
if (potenciaMotor>=4000) {
    return 2;
}
return 0;
`

andrydiaz commented on src/main/java/com/example/springBoot_hibernate_ferrocarriles/service/EquipoDeTraccionService.java:
```diff
@@ -0,0 +1,45 @@
+		this.equipoDeTraccionRepository = equipoDeTraccionRepository;
+	}
+	
+	public List<EquipoDeTraccion> getAllEquipoDeTraccion(){
```
It is a good practice to log information about what the system is doing. A rule of thumb is to use loggers in all your Service class, because in general, they contain the dirty code executing the business rules. Example:
`
public List<EquipoDeTraccion> getAllEquipoDeTraccion(){
log.info("Retrieving all traction equipment");
......
}
`

andrydiaz commented on src/main/java/com/example/springBoot_hibernate_ferrocarriles/service/EquipoDeTraccionService.java:
```diff
@@ -0,0 +1,45 @@
+		return equiposDeTraccion;		
+	}
+	
+	public EquipoDeTraccion getEquipoDeTraccionById(Long idEquipoDeTraccion){
```
Another example of logging:
```java
public EquipoDeTraccion getEquipoDeTraccionById(Long idEquipoDeTraccion){
log.info("Retrieving traction equipment by ID: {}", idEquipoDeTraccion);
...
}
```

andrydiaz replied:
There a lot of loggers out there, but a very easy one is to use a library called Lombock, that among other interesting things provides a log by just annotating the class with `@Slf4j`
Download lombock [here](https://projectlombok.org/download)
Guide to set up your IDE: [this blog](https://www.baeldung.com/lombok-ide)
Anything that you can find about Java in the above blog(baeldung) is interesting and worthy knowing

andrydiaz commented on src/main/java/com/example/springBoot_hibernate_ferrocarriles/service/ItinerarioService.java:
```diff
@@ -0,0 +1,45 @@
+
+@Service
+public class ItinerarioService {
+	@Autowired
```
The latest versions of Spring don't require `@Atowired` annotation anymore. try if it is your case.

andrydiaz commented:
In general good style, variable naming and organization in terms of the project structure. I like it, for somebody who has not coded in a long time, it is outstanding.

lianet89 committed:
Fixing the code

The main changes in the code was:
- Creating a ReportesService class.
- Moving some methods from ReportesControllers class to ReportesService class (That methods  have logic inside).
- Spliting complex methods in smaller methods.
- Modifying the URLs format.
- Deleting constructors of the Controller classes.
- Using loggers in all services class (Lombok and @Slf4j).
NOTE: I am currently using SpringBoot 3.0.8, in wich is neccessary to use @Autowired annotation.
